stab_controller_multi_lqr_node.cpp

```diff
@@ -0,0 +1,416 @@
+/* © Ascento; focusproject 2017/2018; ASL ETH Zurich
+___   _____ _____________   _____________
+/   | / ___// ____/ ____/ | / /_  __/ __  |
+/ /| | \__ \/ /   / __/ /  |/ / / / / / / /
+/ ___ |___/ / /___/ /___/ /|  / / / / /_/ /
+/_/  |_/____/\____/_____/_/ |_/ /_/  \____/
+Author: Lionel Gulich <lgulich@student.ethz.ch> - February 2018
+--------------------------------------------------------------------------------
+Code provided under the GNU General Public Licence GNU GPLv3
+--------------------------------------------------------------------------------
+This code provides the framework for the stabilisation control of the sara
+prototype. It uses multiple lqr controllers derived from different equilibrium
+positions of the system. Between the positions the controllers are interpolated.
+For read in of the k matrix array the matrices should be stored in "/resources"
+as a .csv file and named as "k_matrix_0", "k_matrix_1", ...
+For read in of the imu_offset scaler array the vectors should be stored
+in "/resources" as a .csv file and named as "imu_offset_0", "imu_offset_1", ...
+--------------------------------------------------------------------------------
+TODO: is double necessary in matrices or would float be enough?
+--------------------------------------------------------------------------------
+*/
+
+//Include header, ros and other directories
+#include <ros/ros.h>
+#include "stab_controller_multi_lqr_node.h"
+
+//Set namespace
+namespace ascento {
+
+  //Constructor / Destructor
+  StabControllerMultiLqrNode::StabControllerMultiLqrNode(ros::NodeHandle& nh) : maxon_left(1,"USB1", "EPOS4"), maxon_right(2,"USB0", "EPOS4"), nh(nh) {
+    ROS_INFO("stab_controller_multi_lqr node has started.");
+
+    //Initialise class members
+    Initialise();
+
+    //Initialise Threads for reading position/velocity and writing torques to maxon motors
+    std::thread ReadLeft(&StabControllerMultiLqrNode::ReadLeft,this);
+    std::thread ReadRight(&StabControllerMultiLqrNode::ReadRight,this);
+    std::thread WriteLeft(&StabControllerMultiLqrNode::WriteLeft,this);
+    std::thread WriteRight(&StabControllerMultiLqrNode::WriteRight,this);
+
+    // Start rated loop
+    RateController();
+
+  }
+  StabControllerMultiLqrNode::~StabControllerMultiLqrNode() {
+    maxon_left.MoveWithTorque(0);
+    maxon_right.MoveWithTorque(0);
+  }
+
+  //Setup methods
+  void StabControllerMultiLqrNode::ResetStates() {
+    encoder_offset_left = maxon_left.GetPosition()*2*M_PI/encoder_quadcounts;
```

Comment:
unless you want an integer its good to be in the habit of writing 2.0, to request a double